tests/testthat/test-age_helpers.R

```diff
@@ -0,0 +1,391 @@
+test_that("age_labels() works", {
+
+  # Test simple use case
+  expect_identical(
+    age_labels(c(5, 12, 20, 30)),
+    c("00-04", "05-11", "12-19", "20-29", "30+")
+  )
+
+  # Test superfluous zeros are ignored
+  expect_identical(
+    age_labels(c(0, 5, 12, 20, 30)),
+    c("00-04", "05-11", "12-19", "20-29", "30+")
+  )
+
+  # Test zero-padding works for large ages
+  expect_identical(
+    age_labels(c(5, 12, 20, 30, 130)),
+    c("000-004", "005-011", "012-019", "020-029", "030-129", "130+")
+  )
+})
+
+
+# Create a test data set that contains all birth dates in 2000, 1970, and 1940 (years with, or almost with, leap years)
+test_data <- c(
+  seq.Date(from = as.Date("2000-01-01"), to = as.Date("2000-12-31"), by = "1 day"),
+  seq.Date(from = as.Date("1970-01-01"), to = as.Date("1970-12-31"), by = "1 day"),
+  seq.Date(from = as.Date("1940-01-01"), to = as.Date("1940-12-31"), by = "1 day")
+) |>
+  tibble::tibble(birth_date = _) |>
+  dplyr::mutate(reference_date = birth_date + lubridate::days(40 * dplyr::row_number()))
+
+
+test_that("age_on_date() works for date input", {
+
+  test_date <- as.Date("2024-02-28")
+
+  for (conn in get_test_conns()) {
+
+    # Copy to the remote
+    test_ages <- dplyr::copy_to(conn, test_data, "test_age")
+    SCDB::defer_db_cleanup(test_ages)
+
+    # Compute the reference age using lubridate
+    reference_ages <- test_data |>
+      dplyr::mutate(age = floor(lubridate::interval(.data$birth_date, test_date) / lubridate::years(1)))
+
+
+    # SQLite does not have a precise way to estimate age, so the age computation helper will throw a warning
+    # We here test that the warning is thrown, and filter out the known offending dates from the test
+    if (inherits(conn, "SQLiteConnection")) {
+
+      # Capture and check warning
+      test_ages <- tryCatch(
+        dplyr::mutate(test_ages, "age" = !!age_on_date("birth_date", test_date, conn)),
+        warning = function(w) {
+          expect_match(w$message, "Age computation on SQLite is not precise")
+          return(suppressWarnings(dplyr::mutate(test_ages, "age" = !!age_on_date("birth_date", test_date, conn))))
+        }
+      )
+
+      # Remove offending dates
```

Comment:
Please explain

Reply:
First off, let me reiterate that SQLite does not have a way to compute the difference in years between dates. It stores dates as integers, and we use (date2 - date1) / 365.25 as a estimate for the number of years between the dates. Which works well in many cases.

So, for some dates, we know that SQLite will just not return the correct age for individuals, and we warn the user that this is the case.

The tests however, will just compare the (wrong) ages from SQLite with the reference ages from lubridate, and fail (even though we expect them to fail).

I therefore remove the date that give a wrong age in our test dataset 1970-02-28 (which lands on one of the cases where the age is computed incorrectly).

The tests will now compare all the dates that should give the right age and check that they indeed give the right age. If something else now breaks in our age computation for SQLite, the tests will still capture this. 



However, I could add an additional test, that checks that this date is wrong in SQLite, since we expect it to be wrong.

Reply:
Maybe, et is better to just keep suggesting DuckDB when wanting to use dates in a local file based database.